### DISCUSSION TITLE
collagraph custom QLayout

### DISCUSSION THREAD

**Discussion by jamesdbrock** [fork-tongue/collagraph#105] [General]
Following up on https://github.com/orgs/fork-tongue/discussions/104

I would like use a custom `QLayout` like https://doc.qt.io/qtforpython-6/examples/example_widgets_layouts_flowlayout.html

```python
renderer = PySideRenderer(autoshow=False)
renderer.register("FlowLayout", FlowLayout)

        cg.h(
            "FlowLayout",
            {},
            cg.h(
                "Button",
                {
                    "text": "New Project",
                },
            ),
        )
```
```
TypeError: Specified type '<class 'flowlayout.FlowLayout'>' not a subclass of QWidget
```

Is there a way to register custom `QLayout`s like we register custom `QWidget`s?

**Comment by jamesdbrock** [comment]
I get the feeling that adding `QLayout`s would require changes to https://github.com/fork-tongue/collagraph/blob/master/collagraph/renderers/pyside/objects/widget.py

**Comment by berendkleinhaneveld** [comment]
Currently it is not supported: only the basic QBoxLayout, QGridLayout and QFormLayout are supported through a special `layout` attribute for a widget. You can see how that works here: https://github.com/fork-tongue/collagraph/blob/c23e173b615f90ebdbc56a658be518a2576d3381/collagraph/renderers/pyside/objects/widget.py#L66

An example of that can be seen here: https://github.com/fork-tongue/collagraph/blob/master/examples/layout-example.py
You need to pass a "layout" dict as a prop to the widget, and pass a "type" to select the type of layout to use.

We could add an extra if-else statement in the `widget.set_attribute` call to check whether the layout type name is present in `TYPE_MAPPING` dict. This would make it possible to register a custom layout with a certain name and then use it as a "type" in the "layout" dict:
```python
renderer = PySideRenderer()
renderer.register("flow", FlowLayout)

cg.h(
    "widget",
    {
        "layout": {
            "type": "flow",
            # ... and any other attributes
            # that should be applied to the
            # layout object
        }
    },
    cg.h(
        "Button",
        {
            "text": "New Project",
        },
    ),
)
```

**Reply by berendkleinhaneveld** [reply]
FYI: I've opened #113 which makes it possible to register custom layouts, like so:
```python
renderer = PySideRenderer()
renderer.register_layout("flow", FlowLayout)
...

# or with a decorator
@PySideRenderer.register_layout("flow")
class FlowLayout(QLayout):
    ...
```
Also custom widgets (and insert/remove/set_attr) can be registered with decorators.
